service/matching/matching_engine.go

```diff
@@ -371,60 +371,129 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
+		err := e.forceLoadAllNonRootPartitions(ctx, partition)
+		if err != nil {
+			e.logger.Error("Failed to force load all non-root partitions after root partition was loaded",
+				tag.WorkflowNamespaceID(partition.NamespaceId().String()),
+				tag.WorkflowTaskQueueName(partition.TaskQueue().Name()),
+				tag.WorkflowTaskQueueType(partition.TaskType()),
+				tag.Error(err))
+		}
+	}
+
 	if err = pm.WaitUntilInitialized(ctx); err != nil {
 		e.unloadTaskQueuePartition(pm, unloadCauseInitError)
 		return nil, err
 	}
 	return pm, nil
 }
 
-// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
-// to get new range from DB and create one. This does not block for the task queue to be initialized.
-func (e *matchingEngineImpl) getTaskQueuePartitionManagerNoWait(
+func (e *matchingEngineImpl) forceLoadAllNonRootPartitions(
+	ctx context.Context,
+	rootPar tqid.Partition,
+) error {
+	namespaceId := rootPar.NamespaceId()
+
+	namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(namespaceId)
+	if err != nil {
+		return err
+	}
+
+	nsName := namespaceEntry.Name()
+	taskQueue := rootPar.TaskQueue()
+
+	partitionTotal := e.config.NumTaskqueueReadPartitions(nsName.String(), taskQueue.Name(), taskQueue.TaskType())
+
+	for partitionId := 1; partitionId < partitionTotal; partitionId++ {
+
+		go func() {
+			resp, err := e.matchingRawClient.ForceLoadTaskQueue(ctx, &matchingservice.ForceLoadTaskQueueRequest{
+				NamespaceId: namespaceId.String(),
+				TaskQueuePartition: &taskqueuespb.TaskQueuePartition{
+					TaskQueue:     taskQueue.Name(),
+					TaskQueueType: taskQueue.TaskType(),
+					PartitionId:   &taskqueuespb.TaskQueuePartition_NormalPartitionId{NormalPartitionId: int32(partitionId)},
+				},
+			})
+			if err != nil {
+				e.logger.Error("Failed to force load non-root partition after root partition was loaded",
+					tag.WorkflowNamespaceID(nsName.String()),
+					tag.WorkflowTaskQueueName(taskQueue.Name()),
+					tag.WorkflowTaskQueueType(taskQueue.TaskType()),
+					tag.Error(err))
+				return
+			}
+
+			if !resp.WasUnloaded {
+				// Generally we should be avoiding trying to force load partitions which are already loaded.
+				e.metricsHandler.Counter(metrics.ForceLoadedTaskQueuePartitionUnnecessarilyCounter.Name()).Record(1)
+			}
+
+		}()
+	}
+	return nil
+}
+
+// Returns a taskQueuePartitionManager and a bool indicating if the taskQueue was newly created.
+// by attempting to get new range from DB. This does not block for the task queue to be initialized.
+func (e *matchingEngineImpl) createTaskQueuePartitionManager(
 	partition tqid.Partition,
-	create bool,
 	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
+) (taskQueuePartitionManager, bool, error) {
 	key := partition.Key()
-	e.partitionsLock.RLock()
+
+	e.partitionsLock.Lock()
+	defer e.partitionsLock.Unlock()
+
+	// Check taskQueue doesn't already exist
 	pm, ok := e.partitions[key]
```

Comment:
I don't really like splitting the double-checked lock idiom across two functions. it makes it harder to see the pattern.

Reply:
I would love to hear your thoughts in response to this: I think the split of the functions makes more sense now with both functions being less coupled and more cohesive. I can call either, from many places, and expect that they each manage their own need for locks properly. Said another way, why need to know about the idiom in the first place, except because it was written that way? 

`
GetTQPartitionManager
    lock
         pull from map
    unlock

    if found
        return

   CreateTQPartitionManager
   return


CreateTQPartitionManager
    lock
         pull from map
    if found
        unlock
        return

   build it....
   unlock
   return
`

Mildly related, I hope you laugh at this like I did: Finding this site really helped me find my religion, and begin my crusade for code readability: https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms

Reply:
but that's the whole point of the idiom: these two functions _should_ be coupled and other callers should _not_ be calling createTaskQueuePartitionManager, because it takes an exclusive lock instead of a shared lock and so limits concurrency. also it's just confusing for callers to have two functions that do very similar things.

the point is to have a single entry point that uses the rwlock to most benefit: first lock shared, check, if not there then re-lock exclusive, check again, create if necessary, start. that whole pattern should be encapsulated and callers should just call one thing, with options for whether to create or not, wait or not. (yes, previously the wait or not decision was a different function instead of a flag, but I'd be fine with making it a flag, against the advice of the linter.)

why idioms at all: because when I see an rwlock and a map, I want to check that the rwlock is being used reasonably and should be an rwlock at all (often it shouldn't). if I see the standard double-checked pattern then I'm good. if not, I have to actually look at all the call sites and figure out how often it's expected to be locked shared vs exclusive and think about whether the benefit of concurrent reads outweighs the greater overhead of the rwlock vs a plain lock. the idiom lets me chunk that. and of course this applies to many more concepts than just rwlocks.

Reply:
I express my most sincere consolation for the time you had to spend with C++

Reply:
Hahaha, I doubt my time In C is over, but I am glad I get to focus back on Go for a while.

First, I tried to search the internet for this issue or code case and came up empty handed. Have any good resources or riot inciting blog posts? 

So, if I understand, the idiom addresses the concern that the RWLock would be better off as a typical Lock. This has never struck me as too concerning. If anything, I notice I feel more at ease when I see calls for RLock because there is opportunity for other visits to the resource to not cause contention. 

Thats not to say, I don't follow your larger concern, rather I just see it as "function Y should only be called by function X". I have to say, I don't have a wonderful solution to this problem. Often I feel like I face a trade off. Either 
1. Put code into reasonably sized blocks. Often best done with function to make the boundaries for the code explicit and give the code block a name, and stack frame for debugging and error handling.
2. Build larger functions, but better ensure code is only executed with surrounding code. 

I prefer 1, but don't find it suits all my needs. Often I wish I could have some sort of language enforced "Y must be called by X" but then I imagine the abuse that such a feature would endure and all the arguments about how "its best practice to specify that Y must be called by L, M, N, O, P, Q, R, S, T, U, V, W, or X"

Aaaanyways, I'm just daydreaming while waiting for the functional tests to pass. What code would you prefer to see here? Just the two functions combined into a larger call?

Reply:
O, also, I totally blanked this, but there are two callers for `createTaskQueuePartitionManager` now. The second one is `ForceLoadTaskQueuePartition` I think that primarily justifies the separation of the functions and thus locks.

Reply:
Of course there are competing concerns, or we wouldn't be talking about it.

I think you expressed the tension well. But that "reasonably sized" is holding a lot of weight. It's obviously subjective and depends on the language, surrounding code, the author, etc. The factors it weighs against are also somewhat subjective, of course.

A few thoughts:

- Functions are an abstraction mechanism, and over-abstraction is a thing.
- Functions force your eye to scan the code in a non-linear way, you have to jump to the definition and then jump back to fully understand what's happening.
- Each new function necessarily introduces three extra lines, which decreases the amount of code you can fit on your screen and see at once.
- The general vibe of Go is more transparency and less abstraction, at the cost of verbosity and even duplication.

Reply:
> What code would you prefer to see here? Just the two functions combined into a larger call?

I do prefer how it was before, yeah. You can collapse the NoWait thing since that's not used anymore.

Reply:
Thanks, your 2nd message was particularly insightful. Some great learnings in there, I especially enjoyed the paper linked in the issue that addressed the possible solution to cache contention in RWLocks.


I'm going to focus on the one thought that I think this now hinges on.

> But it shouldn't call it, it should call getTaskQueuePartitionManager with create true, **so that if the partition is already loaded (which is common)**, it hasn't taken an unnecessary exclusive lock.
> 
> Yes, of course this one is too low-volume to matter, it's just the principle that you can't use the wrong code to justify an argument :)

I think this use case is uniquely different. We are building this feature particularly with the assumption that loading a root partition, and hence multiple calls to ForceLoadTaskQueuePartition, assumes that it is more common that the non-root partitions are not loaded. 

In that case, I think [golang issue 17973](https://www.github.com/golang/go/issues/17973) supports the skipping the R(Un)Lock check, as we only add to the cache contention.



With that all said, I think we have thoroughly picked apart the motivations and possible solutions here. If you see this last thought and still prefer the idiom, I'll put it back :)

Reply:
Also, This conversation has been informative and challenging in many ways. Thanks for engaging. This takes me back to undergrad research and writing papers :)

Reply:
I agree with that analysis. But it's almost proving the point about maintainability (both writing and reviewing) that I brought up above: I don't _want_ to do an analysis like this every time I see a call to obtain a tqpm. There should just be one thing to call, no choices, and it should encapsulate all the locking concerns. Putting concerns like that on the caller make reasoning about code more complicated. If it's a very hot path then maybe the extra complication is worth it, but it's clearly not here.

Reply:
Done.